noir_stdlib/src/collections/bounded_vec.nr

```diff
@@ -1,66 +1,271 @@
 use crate::{cmp::Eq, convert::From};
 
+/// A `BoundedVec<T, MaxLen>` is a growable storage similar to a `Vec<T>` except that it
+/// is bounded with a maximum possible length. Unlike `Vec`, `BoundedVec` is not implemented
+/// via slices and thus is not subject to the same restrictions slices are (notably, nested
+/// slices - and thus nested vectors as well - are disallowed).
+/// 
+/// Since a BoundedVec is backed by a normal array under the hood, growing the BoundedVec by
+/// pushing an additional element is also more efficient - the length only needs to be increased
+/// by one.
+/// 
+/// For these reasons `BoundedVec<T, N>` should generally be preferred over `Vec<T>` when there
+/// is a reasonable maximum bound that can be placed on the vector.
+/// 
+/// Example:
+/// 
+/// ```noir
+/// let mut vector: BoundedVec<Field, 10> = BoundedVec::new();
+/// for i in 0..5 {
+///     vector.push(i);
+/// }
+/// assert(vector.len() == 5);
+/// assert(vector.max_len() == 10);
+/// ```
 struct BoundedVec<T, let MaxLen: u32> {
     storage: [T; MaxLen],
     len: u32,
 }
 
 impl<T, let MaxLen: u32> BoundedVec<T, MaxLen> {
+    /// Creates a new, empty vector of length zero.
+    ///
+    /// Since this container is backed by an array internally, it still needs an initial value
+    /// to give each element. To resolve this, each element is zeroed internally. This value
+    /// is guaranteed to be inaccessible unless `get_unchecked` is used.
+    ///
+    /// Example:
+    ///
+    /// ```noir
+    /// let empty_vector: BoundedVec<Field, 10> = BoundedVec::new();
+    /// assert(empty_vector.len() == 0);
+    /// ```
+    ///
+    /// Note that whenever calling `new` the maximum length of the vector should always be specified
+    /// via a type signature:
+    ///
+    /// ```noir
+    /// fn good() -> BoundedVec<Field, 10> {
+    ///     // Ok! MaxLen is specified with a type annotation
+    ///     let v1: BoundedVec<Field, 3> = BoundedVec::new();
+    ///     let v2 = BoundedVec::new();
+    ///
+    ///     // Ok! MaxLen is known from the type of foo's return value
+    ///     v2
+    /// }
+    ///
+    /// fn bad() {
+    ///     let mut v3 = BoundedVec::new();
+    ///
+    ///     // Not Ok! We don't know if v3's MaxLen is at least 1, and the compiler often infers 0 by default.
+    ///     v3.push(5);
+    /// }
```

Comment:
Oh, looks like we should update the docs here. The compiler should now error if a generic is not specified and can't be inferred.

Reply:
Thanks! I updated the `.md` too.

For now I'm leaving code that won't compile in the example code, uncommented. If we eventually can type-check or run this code, it will fail... but I guess then we can have a special directive to say "this shouldn't compile".

Reply:
I made an issue for now to update the outdated docs